src/serializers/type_serializers/simple.rs

```diff
@@ -114,12 +115,15 @@ macro_rules! build_simple_serializer {
                 let py = value.py();
                 match extra.ob_type_lookup.is_type(value, $ob_type) {
                     IsType::Exact => Ok(value.into_py(py)),
-                    IsType::Subclass => match extra.mode {
-                        SerMode::Json => {
-                            let rust_value = value.extract::<$rust_type>()?;
-                            Ok(rust_value.to_object(py))
-                        }
-                        _ => infer_to_python(value, include, exclude, extra),
+                    IsType::Subclass => match extra.check {
+                        SerCheck::Strict => Err(PydanticSerializationUnexpectedValue::new_err(None)),
+                        SerCheck::Lax | SerCheck::None => match extra.mode {
```

Comment:
I think by introducing this you need to implement `retry_with_lax_check` for these serializers.

I think a test which attempts to serialize an int subclass in a `Union[bool, int]` would show this to be necessary.

Reply:
You are right, this test fails due to falling back to `infer_to_python`. I've added `retry_with_lax_check` which is always `false` for bool type (as there should be no bool subtypes in python) and `true` for int type